Adds new API CosmosContainer.readAllItems to retrieve all documents of a logical partition avoiding IO to retrieve query plan

FabianMeiswinkel opened the pull request.

kirankumarkolli commented on sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java:
```diff
@@ -532,6 +534,102 @@ public String getId() {
+    @Beta(Beta.SinceVersion.V4_4_0)
+    public <T> Mono<FeedResponse<T>> readMany(
+        List<Pair<String, PartitionKey>> itemKeyList,
+        String sessionToken,
```
Is Itemizing (or promoting parts of) request option, to limit to applicable ones?

FabianMeiswinkel replied:
Yes - the only relevant one is SessionToken compared to CosmosQueryRequestOptions

FabianMeiswinkel commented:
/azp run java - cosmos - tests

kushagraThapar commented on eng/jacoco-test-coverage/pom.xml:
```diff
@@ -104,12 +104,12 @@
       <groupId>com.azure</groupId>
       <artifactId>azure-cosmos-encryption</artifactId>
-      <version>1.0.0-beta.1</version> <!-- {x-version-update;com.azure:azure-cosmos-encryption;current} -->
+      <version>1.1.0-beta.1</version> <!-- {x-version-update;com.azure:azure-cosmos-encryption;current} -->
```
We don't need to increment this version - since cosmos-encryption is not releasing.

FabianMeiswinkel replied:
Fixed

kushagraThapar commented on eng/jacoco-test-coverage/pom.xml:
```diff
@@ -360,7 +360,7 @@
       <groupId>com.azure</groupId>
       <artifactId>azure-spring-data-cosmos-core</artifactId>
-      <version>3.0.0-beta.2</version> <!-- {x-version-update;com.azure:azure-spring-data-cosmos-core;current} -->
+      <version>3.1.0-beta.1</version> <!-- {x-version-update;com.azure:azure-spring-data-cosmos-core;current} -->
```
Same here, we don't need these version increments.

FabianMeiswinkel replied:
Fixed

FabianMeiswinkel replied:
Fixed

kushagraThapar commented on eng/versioning/version_client.txt:
```diff
@@ -19,9 +19,9 @@ com.azure:azure-core-serializer-json-gson;1.0.0-beta.3;1.0.0-beta.4
-com.azure:azure-cosmos-encryption;1.0.0-beta.1;1.0.0-beta.1
+com.azure:azure-cosmos;4.3.1;4.4.0-beta.1
+com.azure:azure-cosmos-benchmark;4.0.1-beta.1;4.1.0-beta.1
+com.azure:azure-cosmos-encryption;1.1.0-beta.1;1.1.0-beta.1
```
benchmark and encryption should not be incremented. Nor should be spring-data-cosmos modules. 
you can even skip azure-cosmos version increment here, and just do it one time when we are ready to release.

FabianMeiswinkel replied:
Fixed

kushagraThapar commented on eng/versioning/version_client.txt:
```diff
@@ -115,7 +115,7 @@ com.microsoft.azure:spring-cloud-azure-stream-binder-test;1.2.8-beta.1;1.2.8-bet
 com.microsoft.azure:spring-cloud-azure-servicebus-queue-stream-binder;1.2.8-beta.1;1.2.8-beta.1
 com.microsoft.azure:spring-cloud-azure-eventhubs-stream-binder;1.2.8-beta.1;1.2.8-beta.1
-com.microsoft.azure:spring-data-cosmosdb;3.0.0-beta.1;3.0.0-beta.1
+com.microsoft.azure:spring-data-cosmosdb;3.0.0-beta.1;3.1.0-beta.1
```
This is not required either.

kushagraThapar commented on sdk/cosmos/azure-cosmos-encryption/CHANGELOG.md:
```diff
@@ -1,4 +1,4 @@
 # Release History
 
-## 1.0.0-beta.1 (Unreleased)
+## 1.1.0-beta.1 (Unreleased)
```
No need to update this. please leave it as it is.

FabianMeiswinkel replied:
Fixed

kushagraThapar commented on sdk/cosmos/azure-cosmos/CHANGELOG.md:
```diff
@@ -1,7 +1,9 @@
 
-## 4.3.2-beta.3 (Unreleased)
-
+## 4.4.0-beta.1 (Unreleased)
```
Thanks for the changelog update.

moderakh commented on sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java:
```diff
@@ -532,6 +534,102 @@ public String getId() {
+    @Beta(Beta.SinceVersion.V4_4_0)
+    public <T> Mono<FeedResponse<T>> readMany(
+        List<Pair<String, PartitionKey>> itemKeyList,
+        String sessionToken,
```
why do we need to expose session token here?
Did anyone ask for it? if not can we remove this?

FabianMeiswinkel replied:
How would a customer who wants to use Session consistency do the ReadMany in this case?

FabianMeiswinkel replied:
The idea was to start exposing this so that not other customers but Adobe take dependency on implementation package

FabianMeiswinkel replied:
If this needs more discussions I will revert it here and leave it as is to not unblock the Walmart fix on this discusison.

moderakh replied:
for the common session scenario session is resolved through internal sessionContainer. 
I would say passing a session token arg is an advanced use case.

FabianMeiswinkel replied:
I would argue readMany is only necessary in advanced scenarios anyway. What is the harm in exposing sessionToken as parameter? Is the concern that you would prefer a special ReadManyRequestOptions with just SessionToken for now?

FabianMeiswinkel replied:
Or don't want to decide on that model yet?

moderakh replied:
I didn't want to decide on the contract/model. but this is beta anyway. 

so makes sense. let's keep it. resolving the comment.

FabianMeiswinkel commented on sdk/cosmos/azure-cosmos-benchmark/src/main/java/com/azure/cosmos/benchmark/AsyncQueryBenchmark.java:
```diff
@@ -123,6 +123,11 @@ protected void performWorkload(BaseSubscriber<FeedResponse<PojoizedJson>> baseSu
+        } else if (configuration.getOperationType() == Configuration.Operation.ReadAllItemsOfLogicalPartition) {
+
+            int index = r.nextInt(1000);
+            String pk = docsToRead.get(index).getProperty(partitionKey);
```
Fixed for other operation types as well

FabianMeiswinkel commented on sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/util/Beta.java:
```diff
@@ -0,0 +1,91 @@
+        V4_4_0,
+
+        /** v4.5.0 */
+        V4_5_0,
```
Ported this straight from Milis's change - i think his intention was to force that only at least Minor versions (not patch versions) bumps are allowed. I kind of liked that model. Do we feel strongly about this?

moderakh replied:
it just sounds weird that we have constants for future unreleased versions here. I would rather to just have only the version we have and we don't add unneeded version constant here.

However not a big deal. no strong feeling. it is your call, whichever you prefer.
Feel free to resolve the comment if you want to keep it they way it is.

milismsft replied:
In the ARM SDK case since we ship this as a separate library, we do need to capture future versions... Not really needed here but it does not hurt either. It's ok either way.

FabianMeiswinkel replied:
Fixed

FabianMeiswinkel commented on sdk/cosmos/azure-cosmos-encryption/pom.xml:
```diff
@@ -13,7 +13,7 @@ Licensed under the MIT License.
   <groupId>com.azure</groupId>
   <artifactId>azure-cosmos-encryption</artifactId>
-  <version>1.0.0-beta.1</version> <!-- {x-version-update;com.azure:azure-cosmos-encryption;current} -->
+  <version>1.1.0-beta.1</version> <!-- {x-version-update;com.azure:azure-cosmos-encryption;current} -->
```
Fixed already - except for azure-cosmos

moderakh commented on sdk/cosmos/azure-cosmos-benchmark/src/main/java/com/azure/cosmos/benchmark/Configuration.java:
```diff
@@ -163,7 +163,8 @@ public Duration convert(String value) {
         ReadThroughputWithMultipleClients,
-        CtlWorkload;
+        CtlWorkload,
+        ReadAllItemsOfLogicalPartition;
```
please add one line description to "-operation" parameter description, similar to other operations.

FabianMeiswinkel replied:
Fixed

moderakh commented on sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosContainer.java:
```diff
@@ -3,22 +3,30 @@
 import com.azure.cosmos.util.CosmosPagedFlux;
 import com.azure.cosmos.util.CosmosPagedIterable;
 import com.azure.cosmos.util.UtilBridgeInternal;
+import io.micrometer.core.instrument.util.StringUtils;
```
dont' rely on StringUtils from micrometer.

Please use `com.azure.cosmos.implementation.apachecommons.lang.StringUtils` instead.

FabianMeiswinkel replied:
Fixed

moderakh commented on sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxDocumentClientImpl.java:
```diff
@@ -1812,6 +1830,75 @@ public QueryCompatibilityMode getQueryCompatibilityMode() {
+                            ModelBridgeInternal.createQueryRequestOptions(options),
+                            range.getId());
+
+                    return createQuery(collectionLink, querySpec, effectiveOptions, Document.class, ResourceType.Document);
```
partitionSplit handling. 
DocumentProducer which is used internally for cross-partition query is recursively split-proof.

Please verify two things:
1) that this code internally uses DocumentProducer (and hence benefit from DB being split proof.)
2) please make sure if partition split happens before DocumentProducer is instantiated still this is split proof.

FabianMeiswinkel replied:
Fixed it - basically I had to change the code to include the PkRangeIdLookup within the retry policy as suspected. Thanks for catching this!
The Producers are used via the normal QueryexecutionContext factory - so we are good there.

FabianMeiswinkel replied:
Actually I only moved the InvalidPartitionExceptionRetryPolicy up one level - so the change didn't impact partition splits. For partition splits my assumption is right now that readAllDocuments will behave like normal queries.

For queries the relevant PkRangeIds are retrieved form the query plan. The query plan is retrieved from the gateway before creating the QueryExecutionContext which then creates the DocumentProducer. QueryExecutionContext.executeAsync and DocumentProducer are handling partition splits.

For readAllDocuments the PkRangeId is set on CosmosQueryRequestOptions and this value is used by the QueryExecutionContext to create the DocumentProducer etc. Everything else is running through the same pipeline

So if there is a problem, because the pkValue -> pkRangeId resolution happens before creating the QueryExecutionContext/DocumentProducer in readAllDocuments the same problem exists for normal queries because the queryPlan has been also retrieved before creating ExecutionContext/DocumentProducer.

Worth some additional investigation and testing - but Mo and I decided offline that it is ok to do that after releasing the initial Beta version.

moderakh commented on sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxDocumentClientImpl.java:
```diff
@@ -1812,6 +1830,75 @@ public QueryCompatibilityMode getQueryCompatibilityMode() {
     }
 
+    @Override
+    public Flux<FeedResponse<Document>> readAllDocuments(
```
collection stale situation:

delete collection, recreate the collection with the same name.

FabianMeiswinkel replied:
Fixed by moving InvalidPartitionExceptionRetryPolicy one level up

moderakh commented on sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/CosmosItemTest.java:
```diff
@@ -200,6 +200,70 @@ public void queryItemsWithContinuationTokenAndPageSize() throws Exception{
+    @Test(groups = { "simple" }, timeOut = TIMEOUT)
+    public void readAllItemsOfLogicalPartition() throws Exception{
+        String pkValue = UUID.randomUUID().toString();
+        InternalObjectNode properties = getDocumentDefinition(UUID.randomUUID().toString(), pkValue);
```
unfortunately some of our existing tests rely on InternalObjectNode which is an internal class. 
Customers see this and may use this as sample.

Please consider replacing InternalObjectNode with ObjectNode for the newly added tests (as this might get used as sample by end users.)

FabianMeiswinkel replied:
Fixed

FabianMeiswinkel commented:
> added a few more comments. other than these looks good to me.
> 
> * split proofing: please make sure this is split proof.
> * collection stale scenario, please make sure we are handling that
> * please replace InternalObjectNode with ObjectNode or POJO for the new tests. (as the tests will be used by the customer as sample)
> * Please add a comment to ItemOperation class referring to the new place the beta api moved too.

- Addressed "Collection Stale scenario" and "InternalObjectNode -> ObjectNode".

- Added comment to ItemOperation that the methods are deprecated. Did not add "@Deprecated" annotation yet - will do in a subsequent PR to reduce risk of issues in CI etc. blocking this hotfix

- Added comment about partition split above. From my understanding readAllDocuments is "as partition split proof" as query today. Will spend some additional time to double-check and add test coverage in subsequent PR after publishing initial Beta version

FabianMeiswinkel commented:
/azp run java - cosmos - tests

FabianMeiswinkel commented:
For " java - cosmos - tests (Cosmos_CI Session_Single_Region_Tcp_Long)" all tests passed - https://tfsprodcus3.visualstudio.com/A0fb41ef4-5012-48a9-bf39-4ee3de03ee35/internal/_TestManagement/Runs?runId=12901490&_a=runCharts and Maven build failed retreieving some SpotBug maven package - I assume this failure can be ignored.

For "java - cosmos - tests" just marked as failed due to the failure above.

I think it is safe to proceed.
